### ISSUE TITLE
Create Check_Runner interface

### ISSUE THREAD

**Issue by mehulkaklotar** [10up/plugin-check#13]
## Description

The Check Runner interface will be implemented by the specific context runner classes like AJAX runner or CLI runner. This interface will have access to the Checks class and its methods to run checks. The interface will have a `run` method that will be the responsibility of classes to implement to start the runner according to their context.

- [x] https://github.com/10up/plugin-check/issues/14
- [x] https://github.com/10up/plugin-check/issues/15

## Acceptance Criteria

- A new `Check_Runner` interface is created and exists at `includes/Checker`.
- The `Check_Runner` interface contains a public `is_plugin_check()` method.
    - The `is_plugin_check` method will return a boolean.
    - The method will be used to determine if the current request is intended for the plugin checker.
- The `Check_Runner` interface contains a public `prepare()` method.
    - The `prepare()` method will be used to run any universal preparations depending on the checks requested.
- The `Check_Runner` interface contains a public `run()` method.
    - The `run` method returns an instance of the `Check_Results` class.
    - The method will run the requested checks against the plugin context and return the results.

**Comment by felixarntz** [comment]
@jjgrainger Can you outline what the `runner` method will work in its two implementations for CLI and AJAX? Is having a single method sufficient? For example, in the AJAX approach, we'll need to run individual checks per AJAX request. How would the `AJAX_Runner` do that if it only has a single `runner` method that has no context of which checks have been run already or which one should be run next?

I think this interface needs a lot more thought on how it should be implemented, to be able to cover both CLI and AJAX. CLI is a lot simpler, so we should probably look at AJAX to come up with a solid approach. We also need to think about how we can express the common parts between them in separate methods. If there's any functionality that's truly shared, we may benefit from an abstract base class implementing this interface too (not saying that we do, but maybe).

Last but not least, why is the method called `runner`? Methods should typically be action-based, runner is a noun.

**Comment by jjgrainger** [comment]
Thanks @felixarntz 

I've updated the AC's to use a `run` method. I've also created a new issue #63 for the `Abstract_Check_Runner` class that contains the logic to run Checks for a plugin.

**Comment by felixarntz** [comment]
@jjgrainger Some feedback on the ACs: I thought we were going to have the runner classes be responsible for detecting which checks to run (based on either WP-CLI input or AJAX request data). The way the ACs are outlined, this is not the case, as the `run()` method receives the `$checks` and `$check_context`, so that would happen somewhere else. Let me know if you had something else in mind, maybe I'm missing something.

I think we need the `run()` method to (probably?) be parameter-less. We also need to add a `prepare()` method, e.g. for conditionally calling the universal preparations. Last but not least, since the runner class is the overall entry point, I think we should probably also have some sort of `is_plugin_check()` or similar method that returns whether the current WordPress load is to run a plugin check.

**Comment by jjgrainger** [comment]
Thanks @felixarntz 

My thinking so far is that the runner classes would be responsible for preparing, running, cleanup and returning a results object. They wouldn't necessarily need to be aware of the context that they were running in and only need to know the checks to run and the plugin to run those checks against.

We have CLI command class that would be responsible for handling the CLI request arguments and would pass the checks and plugin to the runner instance. The CLI class is then also responsible for handling the results returned and formatting them correctly.

I imagine the same to be true for the AJAX request. We'll have an `Admin_AJAX` class that will handle the AJAX request, pass the checks and plugin to the runner, and then format and return the results in the response. The only difference here to the CLI command is that we are only one check per request, so the state of what checks have been run and are remaining will be passed back and forth from the client and server.

However, reading over your comment I wonder if I'm missing something? You mention that the runner class is the "overall entry point" which I'm no clear on. As I see it the entry points will be the CLI command, where the command class method is called, and the AJAX endpoint handled by a class with a `wp_ajax_{$action}` hook.

You've also mentioned a `is_plugin_check()` method/function to determine if the current load is to run a plugin check and we have had discussions on needing to use the object cache drop-in to initialize the plugin checker early enough. I'm wondering if I'm overlooking something?

Let me know if the above makes sense, maybe we can schedule a call for this week to talk over this in detail?

Thanks

**Comment by felixarntz** [comment]
@jjgrainger Foundationally, I think your rationale makes sense. Having dedicated classes for CLI and AJAX are necessary anyway, and those could pass the necessary arguments to the respective runner. However, we have to consider how all of this is going to happen timing-wise as part of the WordPress load cycle. Part of the logic needs to be already initialized in our `object-cache.php` drop-in, and that is before either the CLI command or the AJAX action would be invoked by how those regularly are handled in WordPress.

So at least for the universal preparations, including exchanging the database table prefix, we will need to have classes that can determine whether the current request is for a WP-CLI plugin check request or for an Admin AJAX plugin check request (regarding your `is_plugin_check()` note). Potentially we even need to parse a few minimum pieces of data needed from CLI arguments / request parameters. This is going to be a bit hacky, since it has to happen outside of WordPress's intended points for when those would be accessible.

Because of the need for that I was thinking that the runner classes would need to have awareness of the context. The `object-cache.php` file could instantiate both runner classes and call their `is_plugin_check()` method. If one of them returns `true`, it would call that runner's `prepare()` method. The `CLI_Command` and `Admin_AJAX` classes would then only have the responsibility to implement the _proper_ interface and hook into the _proper_ actions as needed, and they would call the respective runner's `run()` method.

So TLDR the complexity that we need to think through here is that we need to have an "extra early" entry point for the `object-cache.php` drop-in, which is far before the CLI command or the AJAX callback would actually run - it's even before (the rest of) our plugin loads.

**Comment by jjgrainger** [comment]
Thanks @felixarntz 

This clears things up.

I've updated the AC's for the interface but wanted to expand on things here.

- `is_plugin_check()` - I think we could use the `$_SERVER['argv']` to determine if the request is for the plugin checker CLI and the `$_REQUEST` for AJAX. Therefore we won't need to pass anything to this method as they can use the relevant global to determine the request.
- `prepare()` - Currently, preparations always return a cleanup function but I'm not sure if it's useful here as they need to run towards the end of the process. This is the reasoning for assigning cleanup functions to a protected property to be called later with the `cleanup()` method.
- `run()` - The run method now accepts no parameters as we can use the `prepare()` method to create the plugin context and checks based on the request. These can be stored as properties to be called later.

Let me know if this makes sense and I can start updating the `CLI_Runner` and `AJAX_Runner` AC's

**Comment by felixarntz** [comment]
@jjgrainger Thanks for the updates. I think this all makes sense, except I would say we should rather stick to the convention of having the `prepare()` method return the cleanup closure, instead of including a `cleanup()` method. This makes for better consistency, and we would still be able to run this cleanup later. For example (not sure we'll do exactly that but) we could `add_action( 'shutdown', $cleanup )`.

**Comment by jjgrainger** [comment]
Thanks @felixarntz 

I've updated the AC's to reflect this. Please review and assign back to me.

**Comment by felixarntz** [comment]
@jjgrainger I see you now added another new method `setup()`. Can you clarify why this is needed? Can't we handle this logic without another public method, e.g. in the constructor, or in `prepare()` or `run()`?

I'm also a bit concerned on the ambiguity between `setup()` and `prepare()`, it's a bit unintuitive what the difference between those methods is from the names since both are related to setup/preparation of whatever the main purpose of the class is.

**Comment by jjgrainger** [comment]
@felixarntz The `setup()` method is implemented in #14 and #15 and I had updated the AC's there to reflect this.

While writing up the AC's I noticed there was shared logic between the 2 runner classes. This logic is now contained within #63 

The `setup()` method is used to setup the `Checks` instance and retrieve the `Check` instances based on the request (either CLI or AJAX). Initially I had this logic running in the `prepare()` method as it's required to determine if we are running any `Runtime_Checks` but found if this is moved into it's own method more of the shared logic could be implemented in the abstract class avoiding some duplication.

I'm happy to rename this method to `create_context()` or something similar to avoid ambiguity, or move this logic back into the `prepare()` method where it's required if that makes more sense?

**Comment by felixarntz** [comment]
@jjgrainger I see now where our ideas diverge a little. I agree with pretty much all you're saying, maybe we should call this method `setup_checks()`?

However my main feedback here is that this method does not need to be part of the public interface. It could just be called in `prepare()` like you're saying, so that means it should rather be `protected` and thus only appear in the abstract class, but not the interface. Does that make sense?

**Comment by jjgrainger** [comment]
Thanks @felixarntz I've made the changes here by removing the `setup()` method. I've updated the other issues with the new method name (`setup_checls()`) and marking this a protected method.

**Comment by felixarntz** [comment]
@jjgrainger LGTM! ✅

**Comment by vishalkakadiya** [comment]
@jjgrainger @felixarntz The PR for this is ready for your review. Thank you!
> https://github.com/10up/plugin-check/pull/70
